File: Data Structures & Algorithms/reverse-a-linked-list/submission-2.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def reverseList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        if not head: return None
        if not head.next: return head

        prev = head
        curr = head.next
        head.next = None


        while curr.next:
            next = curr.next
            curr.next = prev
            prev = curr
            curr = next
        
        curr.next = prev
        return curr








        
        

File: Data Structures & Algorithms/remove-node-from-end-of-linked-list/submission-2.cs
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     public int val;
 *     public ListNode next;
 *     public ListNode(int val=0, ListNode next=null) {
 *         this.val = val;
 *         this.next = next;
 *     }
 * }
 */

public class Solution {
            public ListNode RemoveNthFromEnd(ListNode head, int n)
            {
                ListNode dummy = new ListNode();
                dummy.next = head;
                ListNode l = dummy;
                ListNode r = dummy;
                ListNode prev = null;


                for (int i = 0; i < n; i++)
                {
                    r = r.next;
                }

                while (r is not null)
                {
                    prev = l;
                    r = r.next;
                    l = l.next;
                }

                prev.next = l.next;
                l.next = null;

                return dummy.next;

            }
}
